cmd/cerbos/repl/internal/repl.go

```diff
@@ -0,0 +1,415 @@
+// Copyright 2021-2022 Zenauth Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+package internal
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"reflect"
+	"sort"
+	"strings"
+
+	_ "embed"
+
+	participle "github.com/alecthomas/participle/v2"
+	enginev1 "github.com/cerbos/cerbos/api/genpb/cerbos/engine/v1"
+	"github.com/cerbos/cerbos/internal/conditions"
+	"github.com/cerbos/cerbos/internal/outputcolor"
+	"github.com/cerbos/cerbos/internal/printer"
+	"github.com/cerbos/cerbos/internal/printer/colored"
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/checker/decls"
+	"github.com/google/cel-go/common"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+	"github.com/google/cel-go/interpreter"
+	"github.com/peterh/liner"
+	exprpb "google.golang.org/genproto/googleapis/api/expr/v1alpha1"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/proto"
+)
+
+var (
+	//go:embed banner.txt
+	banner string
+	//go:embed help.txt
+	helpText string
+
+	errExit        = errors.New("exit")
+	errInvalidExpr = errors.New("invalid expr")
+
+	listType = reflect.TypeOf([]interface{}{})
+	mapType  = reflect.TypeOf(map[string]interface{}{})
+)
+
+const (
+	commentPrefix   = '#'
+	directivePrefix = ':'
+	prompt          = "-> "
+	secondaryPrompt = "> "
+)
+
+type REPL struct {
+	vars     variables
+	decls    map[string]*exprpb.Decl
+	reader   *liner.State
+	parser   *participle.Parser
+	output   Output
+	toRefVal func(interface{}) ref.Val
+}
+
+func NewREPL(reader *liner.State, output Output) (*REPL, error) {
+	parser, err := NewParser()
+	if err != nil {
+		return nil, err
+	}
+
+	repl := &REPL{
+		reader:   reader,
+		parser:   parser,
+		output:   output,
+		toRefVal: conditions.StdEnv.TypeAdapter().NativeToValue,
+	}
+
+	return repl, repl.reset()
+}
+
+func (r *REPL) Loop() error {
+	r.output.Println(banner)
+	r.output.Println(fmt.Sprintf("Type %s to get help", colored.REPLCmd(":help")))
+	r.output.Println()
+
+	for {
+		input := r.readInput()
+		if input == "" {
+			continue
+		}
+
+		r.reader.AppendHistory(input)
+		if err := r.handleInput(input); err != nil {
+			if errors.Is(err, errExit) {
+				return nil
+			}
+
+			if !errors.Is(err, errInvalidExpr) {
+				r.output.PrintErr("Failed to process input", err)
+			}
+		}
+	}
+}
+
+func (r *REPL) readInput() string {
+	var input strings.Builder
+	currPrompt := prompt
+
+	for {
+		line, err := r.reader.Prompt(currPrompt)
+		if err != nil {
+			if errors.Is(err, io.EOF) || errors.Is(err, liner.ErrPromptAborted) {
+				return ""
+			}
+
+			r.output.PrintErr("Failed to read input", err)
+			return ""
+		}
+
+		line = strings.TrimSpace(line)
+		if line == "" {
+			return input.String()
+		}
+
+		if strings.HasSuffix(line, "\\") {
```

Comment:
Clearly not trivial, but it would be really nice to be able to detect unfinished expressions and avoid a hard requirement to append the trailing `\`, so that you could paste multiline JSON directly into the REPL.

Reply:
You read my mind 😄 

I got bogged down in edge cases like nested strings `"foo\"bar'x'baz\"quux"` while trying to implement that. The Participle parser is not much help here and I was writing a parser by hand at that point. It was taking much longer than I anticipated so I decided to park it for a while and just get the feature out for now.